Enumerable Methods

chasscepts opened the pull request:
We implemented the following methods to behave like their corresponding enumerable methods

- my_each
- my_each_with_index
- my_select
- my_all
- my_any
- my_none
- my_count
- my_map
- my_inject

NOTE: We made a mistake and were committing directly to the develop branch. When we discovered the mistake, we created a new project and copied over our work.

Addis-Belete committed:
implement my_each

Addis-Belete committed:
imolement my_each with index

chasscepts committed:
implement my_select method

Addis-Belete committed:
implement my_all

chasscepts committed:
implement my_any method

Addis-Belete committed:
implement my_none

chasscepts committed:
implement my_count method

Addis-Belete committed:
implement my_map

Addis-Belete committed:
implement my_map

chasscepts committed:
implement my_inject method

Addis-Belete committed:
modify my_map to proc

Addis-Belete committed:
modify my_map to accept both block and proc

chasscepts committed:
fix linter errors

Addis-Belete committed:
update readme file

chasscepts committed:
implement multiply_els to test my_inject method

GzimAsani reviewed (changes requested):
# Status: Changes Required :red_circle: 

### Hello team, great job, your code is clean and overall you did very good so far but we have some errors which will need to be fixed:

### my_each
- [ ] returns an Enumerator if no block is given
     LocalJumpError:
       no block given (yield)
- [ ] when a block is given when self is a range returns the range itself after calling the given block once for each element in self
       expected: 5..50
            got: [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]

### my_each_with_index
- [ ] returns an enumerator if no block is given
     LocalJumpError:
       no block given (yield)
- [ ] when a block is given when self is a range returns the range itself after calling the given block once for each element in self
### my_select
- [ ] returns an enumerator if no block is given
     LocalJumpError:
       no block given (yield)

### my_all
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
     NoMethodError:
       undefined method `nil' for nil:NilClass
- [ ]  when no block or argument is given returns false when one of the collection members are false or nil
     NoMethodError:
       undefined method `nil' for nil:NilClass
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [ ] hen a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
     NoMethodError:
       undefined method `nil' for Numeric:Class
       Did you mean?  nil?
- [ ] when a class is passed as an argument returns false if any of the collection is not a member of such class
- [ ]  when a Regex is passed as an argument returns true if all of the collection matches the Regex 
NoMethodError:
        undefined method `nil' for /d/:Regexp
        Did you mean?  nil?
- [ ] when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [ ] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

### my_any
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

### my_none
- [ ] when no block or argument is given returns true only if none of the collection members is true
- [ ] #my_none? when no block or argument is given returns false only if one of the collection members is true

### others

- [ ] #my_count counts the number of items in enum that are equal to item if an argument is given
- [ ] #my_count counts the number of elements yielding a true value if a block is given
- [ ] #my_count does not mutate the original array
- [ ] #my_map returns a new array with the results of running a given block once for every element in enum.
- [ ] #my_map returns a new array with the results of running a given block once for every element in enum.
- [ ]  #my_map returns a new array with the results of running a given block once for every element in enum::range
- [ ] #my_map returns an Enumerator if no block is given
- [ ] #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
- [ ] #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum
- [ ] #my_map does not mutate the original array
- [ ] #my_inject searches for the longest word in an array of string
- [ ] #my_inject does not mutate the original array
- [ ] #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
- [ ]  #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- [ ] #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block

Please fix the required changes and ask for another code review whenever you're ready

_**If you have any question related to the required changes feel free to text me on slack @GzimAsani** or just add a comment here_

HAPPY CODING :computer: :desktop_computer:

chasscepts committed:
refactor methods to pass failing tests

chasscepts committed:
fix linter errors

chasscepts committed:
fix linter errors

chasscepts commented:
We have refactored the code to pass the failing tests.
All the referenced tests that we were able to reproduce are now passing in our development machines.

bini-i commented on enumerables.rb:
```diff
@@ -0,0 +1,127 @@
+      result
+    end
+    to_enum
+  end
```
### my_select

- [x] returns an array containing all elements of enum for which the given block returns a true value
`array.my_select(&block)`
#### expected
`[1, 3, 2, 0, 0, 1, 2, 2, 0, 2, 3, 1, 0, 2, 2, 1, 2, 3, 0, 3, 0, 3, 3, 1, 0, 2, 3, 2, 1, 2, 2, 3, 1, 1, 1, 3, 1, 1, 3, 3, 2, 0, 3, 0, 3, 1, 2]`
#### got
`#<Enumerator: [8, 1, 3, 8, 7, 6, 2, 4, 0, 7, 0, 1, 2, 7, 7, 2, 8, 0, 4, 7, 2, 4, 3, 1, 4, 5, 0, 5, 2,..., 8, 1, 3, 1, 8, 7, 1, 3, 6, 5, 3, 2, 0, 6, 8, 8, 8, 4, 5, 3, 0, 7, 3, 6, 7, 1, 6, 4, 6, 2, 4]:each>`
- [x] returns an array containing all elements of enum for which the given block returns a true value::range
`range.my_select(&block)`
#### expected
`[]`
#### got
`#<Enumerator: 5..50:each>`

bini-i commented on enumerables.rb:
```diff
@@ -0,0 +1,127 @@
+
+  def my_any?(pattern = nil)
+    if block_given? || pattern.nil?
+      helper = block_given? ? block : proc { |value| value }
```
### my_any

- raises an error when a block is passed `undefined local variable or method 'block' for Array`, 
make sure my_any? method expects a block in its parameters.

bini-i reviewed (changes requested):
# Status: Changes Required :yellow_circle:    
### Hello  team  :wave: 

Thanks for working on the requirements suggested by the previous TSE, but you have remaining tests that are not passing. Please refer to the inline comments below :point_down: .

### Variables used for testing

```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

array = Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } 
block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
words = %w[dog door rod blade] 
range = Range.new(5, 50) 
hash = { a: 1, b: 2, c: 3, d: 4, e: 5 } 
numbers = [1, 2i, 3.14] 
array_clone = array.clone
```

#### If you have any question kindly reach me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (TSE)
> GitHub: [@bini-i](https://github.com/bini-i)
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)

chasscepts committed:
add missing &block to my_any? parameters and explicitly return an array in my_select so it doesn't fall through to return an enum

chasscepts commented:
After the last feedback, we discovered and corrected the following errors in our code

- We did not explicitly return the result when a block is given in my_select method. The execution was falling through to return an enum.
- We did not add an explicit block parameter to our my_any? method.

We have corrected both errors and therefore want to request for a new code review.

bini-i reviewed (approved):
# Status: Approved :heavy_check_mark:   
### Hello  team :wave: 

### Congratulations :tada:  your project meets all the requirements and it is approved.

#### If you have any question kindly reach me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (TSE)
> GitHub: [@bini-i](https://github.com/bini-i)
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)
